### PR TITLE
[Python] Rework internals of object registration

### PR DESCRIPTION
**Pull request by Tishj**
Previously we would create TEMP VIEWs for every registered object.
This is unintuitive behavior and could also silently cause DuckDB to overwrite temporary views created by users.

In <https://github.com/duckdb/duckdb/pull/12161> we cleaned up how replacement scans take place and fix the issues around lifetime of referenced python objects.

This enables us to use this same mechanism for registered objects, everywhere that a replacement scan can be used, registered objects can also be used[1].

### Internal details:
Through a ClientContextState we create a bridge between the DuckDBPyConnection and the PythonReplacementScan.
The `registered_objects` are now stored inside the (new) `PythonContextState`.

`DuckDBPyConnection.register` adds to this map.
`DuckDBPyConnection.unregister` removes from this map.
`PythonReplacementScan` now tries to find a suitable registered object before any frames are scanned.

### Other changes
- `DuckDBPyConnection.from_df` and `DuckDBPyConnection.from_arrow` had weird implementations that were handrolling the replacement, causing a couple issues.
These have been simplified to just do a replacement into a TableRef and a TableFunctionRelation is created from that.

- `Bind(BaseTableRef &ref)` handrolled logic to concatenate catalog+schema+table names, this has been unified with `ParseInfo::QualifierToString`

- The `external_dependencies` from `ClientContext` was removed, this was only used in the python client and was needed because created VIEWs could not contain ExternalDependencies, so the Connection was responsible for keeping the objects that the VIEWs relied on alive - this is no longer necessary because of <https://github.com/duckdb/duckdb/pull/12161>

[1] One place where this isn't true is DESCRIBE, as it only supports VIEWs and TABLEs at the moment, support would need to be added for replacement scans there to enable the behavior we lost here.

### CODE REVIEW

**Comment by Tishj** [comment]
@Mytherin should I add support for replacement scans to DESCRIBE before this gets merged, or are we fine with this regression in behavior for now?

**Comment by Tishj** [comment]
Ah.. seems I didn't test this enough, registered objects do not take precedence over temporary views, which is a different behavior than before.
That is kind of the issue this solves as well though, because previously created temporary views by the same name are currently silently deleted, replaced by the registered object.

Since nobody has made a report of this issue, it's safe to assume nobody is relying on registered objects taking precedence over temporary views. But it is something to keep in mind.

File: src/planner/binder/tableref/bind_basetableref.cpp [hunk 1]
```diff
@@ -184,11 +184,7 @@ unique_ptr<BoundTableRef> Binder::Bind(BaseTableRef &ref) {
 		}
 	}
 	if (!table_or_view) {
-		string table_name = ref.catalog_name;
-		if (!ref.schema_name.empty()) {
-			table_name += (!table_name.empty() ? "." : "") + ref.schema_name;
-		}
-		table_name += (!table_name.empty() ? "." : "") + ref.table_name;
+		string table_name = ParseInfo::QualifierToStringNoQuotes(ref.catalog_name, ref.schema_name, ref.table_name);
```
Comment (Mytherin) [inline src/planner/binder/tableref/bind_basetableref.cpp:187]: This is a work-around for replacement scans not having a schema/catalog - can we instead extend the replacement scans with schema/catalog and remove this code entirely?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! One comment. Let's also look at making DESCRIBE work for replacement scans in a separate PR.

**Comment by Mytherin** [comment]
Thanks! LGTM

**Pull request merged**